sys/net/routing/nhdp/nhdp_address.c

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (C) 2014 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     nhdp
+ * @{
+ *
+ * @file
+ * @brief       Centralized address storage implementation for NHDP
+ *
+ * @author      Fabian Nack <nack@inf.fu-berlin.de>
+ *
+ * @}
+ */
+
+#include "mutex.h"
+#include "utlist.h"
+
+#include "nhdp.h"
+#include "nhdp_address.h"
+
+/* Internal variables */
+static mutex_t mtx_addr_access = MUTEX_INIT;
+static nhdp_addr_t *nhdp_addr_db_head = NULL;
+
+
+/*---------------------------------------------------------------------------*
+ *                      Centralized Address Storage API                      *
+ *---------------------------------------------------------------------------*/
+
+nhdp_addr_t* nhdp_addr_db_get_address(uint8_t *addr, size_t addr_size, uint8_t addr_type)
+{
+    nhdp_addr_t *addr_elt;
+
+    mutex_lock(&mtx_addr_access);
+
+    LL_FOREACH(nhdp_addr_db_head, addr_elt) {
+        if ((addr_elt->addr_size == addr_size) && (addr_elt->addr_type == addr_type)) {
+            if (memcmp(addr_elt->addr, addr, addr_size) == 0) {
+                /* Found a matching entry */
+                break;
+            }
+        }
+    }
+
+    if (!addr_elt) {
+        /* No matching entry, create a new one */
+        addr_elt = (nhdp_addr_t*) malloc(sizeof(nhdp_addr_t));
+        if (!addr_elt) {
+            /* Insufficient memory */
+            return NULL;
+        }
+
+        /* Allocate space for the address */
+        addr_elt->addr = (uint8_t*) malloc(addr_size * sizeof(uint8_t));
+        if (!addr_elt->addr) {
+            /* Insufficient memory */
+            free(addr_elt);
+            return NULL;
+        }
+
+        memcpy(addr_elt->addr, addr, addr_size);
+        addr_elt->addr_size = addr_size;
+        addr_elt->addr_type = addr_type;
+        addr_elt->usg_count = 0;
+        addr_elt->in_tmp_table = NHDP_ADDR_TMP_NONE;
+        LL_PREPEND(nhdp_addr_db_head, addr_elt);
+    }
+
+    nhdp_increment_addr_usage(addr_elt);
+
+    mutex_unlock(&mtx_addr_access);
+    
+    return addr_elt;
+}
+
+void nhdp_increment_addr_usage(nhdp_addr_t *addr)
+{
+    addr->usg_count += 1;
```

Comment:
Clean and better understandable code. Does it make sense to make this inline, or would you vote for removing this function?